internal/provider/resource_gitlab_project_milestone.go

```diff
@@ -0,0 +1,187 @@
+		ReadContext:   resourceGitlabProjectMilestoneRead,
+		UpdateContext: resourceGitlabProjectMilestoneUpdate,
+		DeleteContext: resourceGitlabProjectMilestoneDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: gitlabProjectMilestoneGetSchema(),
+	}
+})
+
+func resourceGitlabProjectMilestoneCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	title := d.Get("title").(string)
+
+	options := &gitlab.CreateMilestoneOptions{
+		Title: &title,
+	}
+	if description, ok := d.GetOk("description"); ok {
+		options.Description = gitlab.String(description.(string))
+	}
+	if startDate, ok := d.GetOk("start_date"); ok {
+		parsedStartDate, err := parseISO8601Date(startDate.(string))
+		if err != nil {
+			return diag.Errorf("Failed to parse start_date: %s. %v", startDate.(string), err)
+		}
+		options.StartDate = parsedStartDate
+	}
+	if dueDate, ok := d.GetOk("due_date"); ok {
+		parsedDueDate, err := parseISO8601Date(dueDate.(string))
+		if err != nil {
+			return diag.Errorf("Failed to parse due_date: %s. %v", dueDate.(string), err)
+		}
+		options.DueDate = parsedDueDate
+	}
+
+	log.Printf("[DEBUG] create gitlab milestone in project %s with title %s", project, title)
+	milestone, resp, err := client.Milestones.CreateMilestone(project, options, gitlab.WithContext(ctx))
+	if err != nil {
+		log.Printf("[WARN] failed to create gitlab milestone in project %s with title %s (response %v)", project, title, resp)
+		return diag.FromErr(err)
+	}
+	d.SetId(resourceGitLabProjectMilestoneBuildId(project, milestone.ID))
+
+	updateOptions := gitlab.UpdateMilestoneOptions{}
+	if stateEvent, ok := d.GetOk("state"); ok {
+		updateOptions.StateEvent = gitlab.String(milestoneStateToStateEvent[stateEvent.(string)])
+	}
+	if updateOptions != (gitlab.UpdateMilestoneOptions{}) {
+		_, _, err := client.Milestones.UpdateMilestone(project, milestone.ID, &updateOptions, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.Errorf("Failed to update milestone ID %d in project %s right after creation: %v", milestone.ID, project, err)
+		}
+	}
+
+	return resourceGitlabProjectMilestoneRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectMilestoneRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project, milestoneID, err := resourceGitLabProjectMilestoneParseId(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] read gitlab milestone in project %s with ID %d", project, milestoneID)
+	milestone, resp, err := client.Milestones.GetMilestone(project, milestoneID, gitlab.WithContext(ctx))
+	if err != nil {
+		if is404(err) {
+			log.Printf("[WARN] recieved 404 for gitlab milestone ID %d in project %s, removing from state", milestoneID, project)
+			d.SetId("")
+			return diag.FromErr(err)
```

Comment:
When reading a resource and it doesn't exist it's actually not an error per se, but we want to remove it from state instead. That's why we return `nil` instead of an error and just clear the id (`d.SetId("")`):

```suggestion
			return nil
```

Reply:
I'm agree with you. Stange that I added this error)